feature/s3/transfermanager/api_op_GetObject.go

```diff
@@ -0,0 +1,1106 @@
+//	var wg sync.WaitGroup
+//	wg.Add(1)
+//
+// // You must read from the r in a separate goroutine to drive getter to get all parts.
+//
+//	go func() {
+//		defer wg.Done()
+//		_, err := io.Copy(file, r)
+//		if err != nil {
+//			log.Fatal("error when writing to local file: ", err)
+//		}
+//	}()
+//
+//	out, err := svc.GetObject(context.Background(), &transfermanager.GetObjectInput{
+//		Bucket: "your-bucket",
+//		Key:    "your-key",
+//		Reader: r,
+//	})
+//
+//	// must wait for r.Read() to finish
+//	wg.Wait()
+//	if err != nil {
+//		log.Fatal("error when downloading file: ", err)
+//	}
+func (c *Client) GetObject(ctx context.Context, input *GetObjectInput, opts ...func(*Options)) (*GetObjectOutput, error) {
+	i := getter{in: input, options: c.options.Copy(), r: input.Reader}
+	for _, opt := range opts {
+		opt(&i.options)
+	}
+
+	return i.get(ctx)
+}
+
+type getter struct {
+	options Options
+	in      *GetObjectInput
+	out     *GetObjectOutput
+	w       *types.WriteAtBuffer
+	r       *ConcurrentReader
+
+	wg sync.WaitGroup
+	m  sync.Mutex
+
+	offset     int64
+	pos        int64
+	totalBytes int64
+	written    int64
+
+	err error
+}
+
+func (g *getter) get(ctx context.Context) (out *GetObjectOutput, err error) {
+	if err := g.init(ctx); err != nil {
+		return nil, fmt.Errorf("unable to initialize download: %w", err)
+	}
+
+	clientOptions := []func(*s3.Options){
+		func(o *s3.Options) {
+			o.APIOptions = append(o.APIOptions,
+				middleware.AddSDKAgentKey(middleware.FeatureMetadata, userAgentKey),
+				addFeatureUserAgent,
+			)
+		}}
+
+	defer close(g.r.ch)
+	if g.in.PartNumber > 0 {
+		return g.singleDownload(ctx, clientOptions...)
+	}
+
+	if g.options.GetObjectType == types.GetObjectParts {
+		if g.in.Range != "" {
+			return g.singleDownload(ctx, clientOptions...)
+		}
+		// must know the part size before creating stream reader
+		out, err := g.options.S3.HeadObject(ctx, &s3.HeadObjectInput{
+			Bucket:     aws.String(g.in.Bucket),
+			Key:        aws.String(g.in.Key),
+			PartNumber: aws.Int32(1),
+		}, clientOptions...)
+		if err != nil {
+			g.r.setErr(err)
+			return nil, err
+		}
+
+		partsCount := int32(math.Max(float64(aws.ToInt32(out.PartsCount)), 1))
+		partSize := int64(math.Max(float64(aws.ToInt64(out.ContentLength)), 1))
+		sectionParts := int32(math.Max(1, float64(g.options.GetBufferSize/partSize)))
+		capacity := sectionParts
+		g.r.setPartSize(partSize)
+		g.r.setCapacity(int32(math.Min(float64(capacity), float64(partsCount))))
+		g.r.setPartsCount(partsCount)
+
+		ch := make(chan getChunk, g.options.Concurrency)
+		for i := 0; i < g.options.Concurrency; i++ {
+			g.wg.Add(1)
+			go g.downloadPart(ctx, ch, clientOptions...)
+		}
+
+		var i int32
+		for i < partsCount {
+			if g.getErr() != nil {
+				break
+			}
+
+			if g.r.getRead() == capacity {
+				capacity = int32(math.Min(float64(capacity+sectionParts), float64(partsCount)))
+				g.r.setCapacity(capacity)
+			}
+
+			if i == capacity {
```

Comment:
if we hit this case, wouldn't we deadlock if `g.r.getRead() != capacity`?

Reply:
`g.r.read` would be increased by reader in another goroutine, that's why we continue the loop to re-check `g.r.getRead()` at line 668